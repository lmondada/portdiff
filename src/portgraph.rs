use std::collections::BTreeMap;

use crate::{
    port::{BoundPort, EdgeEnd, Site},
    Graph, PortDiff,
};

use itertools::Itertools;
use pg::{LinkMut, LinkView, PortGraph, PortMut, PortView};
use portgraph as pg;
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize)]
pub struct PortgraphEdge {
    outgoing: u16,
    node: pg::NodeIndex,
}

impl PortgraphEdge {
    pub fn new(node: pg::NodeIndex, outgoing: u16) -> Self {
        Self { node, outgoing }
    }

    pub fn out_node(&self) -> pg::NodeIndex {
        self.node
    }

    pub fn out_offset(&self) -> pg::PortOffset {
        pg::PortOffset::Outgoing(self.outgoing)
    }

    pub fn in_node(&self, g: &PortGraph) -> pg::NodeIndex {
        let port = g.port_index(self.out_node(), self.out_offset()).unwrap();
        let in_port = g.port_link(port).unwrap();
        g.port_node(in_port).unwrap()
    }

    pub fn in_offset(&self, g: &PortGraph) -> pg::PortOffset {
        let port = g.port_index(self.out_node(), self.out_offset()).unwrap();
        let in_port = g.port_link(port).unwrap();
        g.port_offset(in_port).unwrap()
    }
}

impl TryFrom<(pg::NodeIndex, pg::PortOffset)> for PortgraphEdge {
    type Error = &'static str;

    fn try_from((node, port): (pg::NodeIndex, pg::PortOffset)) -> Result<Self, Self::Error> {
        let outgoing = match port {
            pg::PortOffset::Outgoing(out) => out,
            _ => return Err("Invalid edge"),
        };
        Ok(Self { outgoing, node })
    }
}

impl Graph for pg::PortGraph {
    type Node = pg::NodeIndex;

    type Edge = PortgraphEdge; // Using outgoing port offset as edge indices

    type PortLabel = pg::PortOffset;

    fn nodes_iter(&self) -> impl Iterator<Item = Self::Node> + '_ {
        PortView::nodes_iter(self)
    }

    fn edges_iter(&self) -> impl Iterator<Item = Self::Edge> + '_ {
        PortView::nodes_iter(self)
            .flat_map(|node| self.links(node, pg::Direction::Outgoing))
            .map(|(outp, _)| {
                PortgraphEdge::try_from((
                    self.port_node(outp).unwrap(),
                    self.port_offset(outp).unwrap(),
                ))
                .unwrap()
            })
    }

    fn get_port_site(
        &self,
        BoundPort { edge, end }: BoundPort<Self::Edge>,
    ) -> Site<Self::Node, Self::PortLabel> {
        let left = self.output(edge.node, edge.outgoing as usize).unwrap();
        let (left, right) = self.port_links(left).exactly_one().unwrap();
        let port_index = match end {
            EdgeEnd::Left => left,
            EdgeEnd::Right => right,
        };
        Site {
            node: self.port_node(port_index).unwrap(),
            port: self.port_offset(port_index).unwrap(),
        }
    }

    fn get_bound_ports(
        &self,
        unbound_port: Site<Self::Node, Self::PortLabel>,
    ) -> impl Iterator<Item = BoundPort<Self::Edge>> + '_ {
        let port_index = self.port_index(unbound_port.node, unbound_port.port);
        port_index.into_iter().flat_map(|port| {
            self.port_links(port).map(|(src, tgt)| {
                match self.port_offset(src).unwrap().direction() {
                    pg::Direction::Incoming => {
                        let edge = PortgraphEdge::try_from((
                            self.port_node(tgt).unwrap(),
                            self.port_offset(tgt).unwrap(),
                        ))
                        .unwrap();
                        let end = EdgeEnd::Right;
                        BoundPort { edge, end }
                    }
                    pg::Direction::Outgoing => {
                        let edge = PortgraphEdge::try_from((
                            self.port_node(src).unwrap(),
                            self.port_offset(src).unwrap(),
                        ))
                        .unwrap();
                        let end = EdgeEnd::Left;
                        BoundPort { edge, end }
                    }
                }
            })
        })
    }

    fn get_sites(
        &self,
        node: Self::Node,
    ) -> impl Iterator<Item = Site<Self::Node, Self::PortLabel>> + '_ {
        self.all_port_offsets(node)
            .map(move |port| Site { node, port })
    }

    fn link_sites(
        &mut self,
        left: Site<Self::Node, Self::PortLabel>,
        right: Site<Self::Node, Self::PortLabel>,
    ) -> (BoundPort<Self::Edge>, BoundPort<Self::Edge>) {
        let (outport, _) = self
            .link_offsets(left.node, left.port, right.node, right.port)
            .unwrap();
        let edge = (
            self.port_node(outport).unwrap(),
            self.port_offset(outport).unwrap(),
        )
            .try_into()
            .unwrap();
        (
            BoundPort {
                edge,
                end: EdgeEnd::Left,
            },
            BoundPort {
                edge,
                end: EdgeEnd::Right,
            },
        )
    }

    fn add_subgraph(
        &mut self,
        graph: &Self,
        nodes: &std::collections::BTreeSet<Self::Node>,
    ) -> std::collections::BTreeMap<Self::Node, Self::Node> {
        println!("Adding subgraph with {} nodes", nodes.len());
        let mut nodes_map = BTreeMap::new();
        println!("N nodes in self: {}", self.node_count());
        for node in nodes {
            let new_node = self.add_node(0, 0);
            nodes_map.insert(*node, new_node);
        }
        println!(
            "N nodes in self after adding subgraph: {}",
            self.node_count()
        );

        // Add every port in `graph` to `self`
        for (&node, &self_node) in &nodes_map {
            self.set_num_ports(
                self_node,
                graph.num_inputs(node),
                graph.num_outputs(node),
                |_, _| {},
            );

            // Add all outgoing edges of `node` with target in `nodes`.
            for port in graph.all_ports(node) {
                let offset = graph.port_offset(port).unwrap();
                for (_, other_port) in graph.port_links(port) {
                    let other_node = graph.port_node(other_port).unwrap();
                    let Some(other_self_node) = nodes_map.get(&other_node) else {
                        // Ignore edges not in induced subgraph
                        continue;
                    };
                    let other_offset = graph.port_offset(other_port).unwrap();
                    if (other_node, other_offset) <= (node, offset) {
                        // By only adding the edge when other is smaller, we
                        // i) avoid duplicating edges and ii) we know that the
                        // other ports have already been resized.
                        self.link_offsets(self_node, offset, *other_self_node, other_offset)
                            .unwrap();
                    }
                }
            }
        }
        nodes_map
    }
}

impl PortDiff<PortGraph> {
    pub fn nodes(&self) -> impl Iterator<Item = pg::NodeIndex> + '_ {
        PortView::nodes_iter(self.graph())
    }
}
