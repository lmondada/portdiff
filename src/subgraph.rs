use std::collections::BTreeSet;
use std::fmt::Debug;

use crate::{
    port::{BoundPort, EdgeEnd},
    Graph,
};

use derive_where::derive_where;
use serde::{Deserialize, Serialize};

#[derive_where(Clone, Default; G: Graph)]
#[derive_where(Debug; G: Graph, G::Node: Debug, G::Edge: Debug)]
#[derive(Serialize, Deserialize)]
#[serde(bound(
    serialize = "G::Node: Serialize, G::Edge: Serialize",
    deserialize = "G::Node: Deserialize<'de>, G::Edge: Deserialize<'de>"
))]
pub struct Subgraph<G: Graph> {
    nodes: BTreeSet<G::Node>,
    edges: BTreeSet<G::Edge>,
}

impl<G: Graph> Subgraph<G> {
    pub fn new(graph: &G, nodes: BTreeSet<G::Node>, edges: BTreeSet<G::Edge>) -> Self {
        assert!(incident_edges(graph, &nodes, &edges));
        Self { nodes, edges }
    }

    pub fn nodes(&self) -> &BTreeSet<G::Node> {
        &self.nodes
    }

    pub fn edges(&self) -> &BTreeSet<G::Edge> {
        &self.edges
    }

    pub fn boundary<'a>(&'a self, graph: &'a G) -> impl Iterator<Item = BoundPort<G::Edge>> + 'a {
        self.nodes.iter().flat_map(move |&n| {
            let ports = graph
                .get_sites(n)
                .flat_map(|site| graph.get_bound_ports(site));
            ports.filter(move |p| !self.edges.contains(&p.edge))
        })
    }
}

fn incident_edges<G: Graph>(
    graph: &G,
    nodes: &BTreeSet<G::Node>,
    edges: &BTreeSet<G::Edge>,
) -> bool {
    for &e in edges {
        let n1 = graph.incident_node(e, EdgeEnd::Left);
        let n2 = graph.incident_node(e, EdgeEnd::Right);
        if !nodes.contains(&n1) || !nodes.contains(&n2) {
            return false;
        }
    }
    true
}

#[cfg(feature = "portgraph")]
#[cfg(test)]
mod tests {
    use itertools::Itertools;
    use portgraph::{LinkMut, NodeIndex, PortGraph, PortMut};
    use rstest::{fixture, rstest};

    use crate::portgraph::PortgraphEdge;

    use super::Subgraph;

    #[fixture]
    fn graph() -> PortGraph {
        let mut graph = PortGraph::new();

        let n0 = graph.add_node(0, 3);
        let n1 = graph.add_node(3, 1);
        let n2 = graph.add_node(1, 3);
        let n3 = graph.add_node(3, 0);

        for i in 0..3 {
            graph.link_nodes(n0, i, n1, i).unwrap();
            graph.link_nodes(n2, i, n3, i).unwrap();
        }
        graph.link_nodes(n1, 0, n2, 0).unwrap();
        graph
    }

    #[rstest]
    fn test_boundary_no_edge(graph: PortGraph) {
        let n1 = NodeIndex::new(1);
        let n2 = NodeIndex::new(2);
        let sub = Subgraph::new(
            &graph,
            [n1, n2].into_iter().collect(),
            [].into_iter().collect(),
        );
        let boundary = sub.boundary(&graph).collect_vec();
        assert_eq!(boundary.len(), 8);
        insta::assert_debug_snapshot!(boundary);
    }

    #[rstest]
    fn test_boundary_edge(graph: PortGraph) {
        let n1 = NodeIndex::new(1);
        let n2 = NodeIndex::new(2);
        let sub = Subgraph::new(
            &graph,
            [n1, n2].into_iter().collect(),
            [PortgraphEdge::new(n1, 0)].into_iter().collect(),
        );
        let boundary = sub.boundary(&graph).collect_vec();
        assert_eq!(boundary.len(), 6);
        insta::assert_debug_snapshot!(boundary);
    }
}
