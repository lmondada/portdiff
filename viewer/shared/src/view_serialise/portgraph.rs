//! A wrapper around PortGraph for ReactFlow

use portdiff::Graph;
use portgraph::{self as pg, PortGraph};
use serde::{Deserialize, Serialize};

use super::ViewSerialise;

impl ViewSerialise for PortGraph {
    fn graph_type(&self) -> &'static str {
        "portgraph"
    }

    fn to_json(&self) -> String {
        serde_json::to_string(&RFGraph::from(self)).unwrap()
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RFNode {
    pub id: String,
    pub num_in_handles: u32,
    pub num_out_handles: u32,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RFEdge {
    pub source: String,
    pub source_handle: u32,
    pub target: String,
    pub target_handle: u32,
}

#[derive(Debug, Default, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RFGraph {
    pub nodes: Vec<RFNode>,
    pub edges: Vec<RFEdge>,
}

impl RFGraph {
    pub fn new() -> Self {
        Default::default()
    }
}

impl<'r> From<&'r pg::PortGraph> for RFGraph {
    fn from(value: &'r pg::PortGraph) -> Self {
        let nodes = value.nodes_iter().map(|n| RFNode {
            id: format!("{:?}", n),
            num_in_handles: <pg::PortGraph as pg::PortView>::num_inputs(value, n) as u32,
            num_out_handles: <pg::PortGraph as pg::PortView>::num_outputs(value, n) as u32,
        });
        let edges = value.edges_iter().map(|e| RFEdge {
            source: format!("{:?}", e.out_node()),
            source_handle: e.out_offset().index() as u32,
            target: format!("{:?}", e.in_node(value)),
            target_handle: e.in_offset(value).index() as u32,
        });
        RFGraph {
            nodes: nodes.collect(),
            edges: edges.collect(),
        }
    }
}
