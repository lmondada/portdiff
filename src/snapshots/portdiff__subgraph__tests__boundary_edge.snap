---
source: src/subgraph.rs
expression: boundary
---
[
    BoundPort {
        edge: PortgraphEdge {
            outgoing: 0,
            node: NodeIndex(0),
        },
        end: Right,
    },
    BoundPort {
        edge: PortgraphEdge {
            outgoing: 1,
            node: NodeIndex(0),
        },
        end: Right,
    },
    BoundPort {
        edge: PortgraphEdge {
            outgoing: 2,
            node: NodeIndex(0),
        },
        end: Right,
    },
    BoundPort {
        edge: PortgraphEdge {
            outgoing: 0,
            node: NodeIndex(2),
        },
        end: Left,
    },
    BoundPort {
        edge: PortgraphEdge {
            outgoing: 1,
            node: NodeIndex(2),
        },
        end: Left,
    },
    BoundPort {
        edge: PortgraphEdge {
            outgoing: 2,
            node: NodeIndex(2),
        },
        end: Left,
    },
]
